### PR TITLE
Various optimizations in triangle, edge and surface loss calculations 

### PR DESCRIPTION
**Pull request by gennarinoos**
- Vectorized most operations to reduce loops
- Replaced the per-triangle collision checks with batch operations
- Improved memory efficiency by reducing intermediate tensor allocations
- Optimized adjacency checking using vectorized operations
- Used squared distances instead of cdist for faster neighbor finding

### CODE REVIEW

**Review by martinnormark — APPROVED** [review]
I have checked the logic, more against the paper than my code tbh. Only found the suggested normalisation factor - feel free to dismiss if you think it should not be added.

Otherwise I'm mostly in doubt about the memory management with `del` statements being too early?

File: src/neural_mesh_simplification/losses/overlapping_triangles_loss.py [hunk 1]
```diff
@@ -121,51 +134,79 @@ def calculate_overlap_loss(
...
+            # Calculate areas using cross product
+            cross_prod = torch.linalg.cross(
+                relevant_v1 - relevant_v0,
+                relevant_v2 - relevant_v0
+            )
+            areas[where_inside] = 0.5 * torch.norm(cross_prod, dim=1)
+
+            del cross_prod, relevant_v0, relevant_v1, relevant_v2
+
+        del v0, v1, v2, inside_mask, source_mask
```
[the remark this answers is not in the document]
Comment (gennarinoos) [reply src/neural_mesh_simplification/losses/overlapping_triangles_loss.py:207]: Good question. empty_cache is very aggressive. I was doing it at each epoch, and during my testing it was slowing things down because of deallocation and re-allocation.
We can re-test it once we are done with lower-level implementation.
I don't believe that affects the backprop, but I'm only saying that because I haven't seen any error. How else would I know if it does?

File: src/neural_mesh_simplification/losses/edge_crossing_loss.py [hunk 2]
```diff
@@ -69,79 +69,32 @@ def find_nearest_triangles(
...
         return crossings
 
     def calculate_loss(
-            self, crossings: torch.Tensor, face_probs: torch.Tensor
+        self, crossings: torch.Tensor, face_probs: torch.Tensor
     ) -> torch.Tensor:
         # Weighted sum of crossings by triangle probabilities
-        loss = torch.sum(face_probs * crossings)
-        return loss
+        return torch.sum(face_probs * crossings, dtype=torch.float32)
```
[the remark this answers is not in the document]
Comment (gennarinoos) [reply src/neural_mesh_simplification/losses/edge_crossing_loss.py:100]: Is this something I removed or something for which you just noticed a discrepancy?

[the remark this answers is not in the document]
Comment (martinnormark) [reply src/neural_mesh_simplification/losses/edge_crossing_loss.py:100]: Something I noticed reviewing it against the paper - we can skip it for now and adjust later.

File: src/neural_mesh_simplification/losses/overlapping_triangles_loss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinnormark) [reply src/neural_mesh_simplification/losses/overlapping_triangles_loss.py:207]: Could be checked with `print(v0.requires_grad)` - if `True` deleting it could break backprop as it's part of computation graph. Fine to check later, it would break if it was a problem.

File: src/neural_mesh_simplification/losses/edge_crossing_loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gennarinoos) [reply src/neural_mesh_simplification/losses/edge_crossing_loss.py:100]: Cool, I'll issue a new PR for this

File: src/neural_mesh_simplification/losses/overlapping_triangles_loss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinnormark) [reply src/neural_mesh_simplification/losses/overlapping_triangles_loss.py:207]: old post, but seems to backup the aggressiveness of empty_cache: https://discuss.pytorch.org/t/why-does-torch-cuda-empty-cache-make-the-gpu-utilization-near-0-and-slow-down-the-training-time/65196

File: src/neural_mesh_simplification/losses/edge_crossing_loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gennarinoos) [reply src/neural_mesh_simplification/losses/edge_crossing_loss.py:100]: Addressed in https://github.com/martinnormark/neural-mesh-simplification/pull/44
